File: frontend/src/types/notification-settings.ts
/**
 * Notification Settings Types
 *
 * Type definitions for user notification preferences.
 * Matches backend API schema for notification settings.
 */

/**
 * User notification settings
 * Controls email notifications, daily digests, reminders, etc.
 */
export interface NotificationSettings {
  /**
   * Unique identifier
   */
  id: string

  /**
   * User ID (owner of these settings)
   */
  user_id: string

  /**
   * Workspace ID (settings are workspace-scoped)
   */
  workspace_id: string

  /**
   * Master toggle: Enable all email notifications
   * When false, all other email notifications are disabled
   */
  email_enabled: boolean

  /**
   * Event Notifications
   */

  /**
   * Notify when a new appointment is booked
   */
  notify_appointment_booked: boolean

  /**
   * Notify when an appointment is cancelled
   */
  notify_appointment_cancelled: boolean

  /**
   * Notify when an appointment is rescheduled
   */
  notify_appointment_rescheduled: boolean

  /**
   * Notify when a client confirms their appointment
   */
  notify_appointment_confirmed: boolean

  /**
   * Daily Digest
   */

  /**
   * Enable daily digest email
   */
  digest_enabled: boolean

  /**
   * Time to send daily digest (HH:MM format, 24-hour)
   * Example: "08:00", "17:30"
   * Null if digest is disabled
   */
  digest_time: string | null

  /**
   * Skip digest on weekends (Saturday and Sunday)
   */
  digest_skip_weekends: boolean

  /**
   * Appointment Reminders
   */

  /**
   * Enable appointment reminder emails
   */
  reminder_enabled: boolean

  /**
   * Minutes before appointment to send reminder
   * Valid values: 15, 30, 60, 120, 1440 (1 day)
   * Null if reminders disabled
   */
  reminder_minutes: number | null

  /**
   * Session Notes Reminders
   */

  /**
   * Enable reminder to complete session notes
   */
  notes_reminder_enabled: boolean

  /**
   * Time to send session notes reminder (HH:MM format, 24-hour)
   * Null if notes reminders disabled
   */
  notes_reminder_time: string | null

  /**
   * Extended settings (for future features)
   * Stored as JSON, can contain any additional settings
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  extended_settings: Record<string, any>

  /**
   * Timestamps
   */
  created_at: string
  updated_at: string
}

/**
 * Notification settings update request
 * All fields are optional (partial updates supported)
 */
export interface NotificationSettingsUpdate {
  email_enabled?: boolean
  notify_appointment_booked?: boolean
  notify_appointment_cancelled?: boolean
  notify_appointment_rescheduled?: boolean
  notify_appointment_confirmed?: boolean
  digest_enabled?: boolean
  digest_time?: string | null
  digest_skip_weekends?: boolean
  reminder_enabled?: boolean
  reminder_minutes?: number | null
  notes_reminder_enabled?: boolean
  notes_reminder_time?: string | null
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  extended_settings?: Record<string, any>
}

/**
 * Valid reminder minute options
 */
export const REMINDER_MINUTE_OPTIONS = [
  { value: 15, label: '15 minutes before' },
  { value: 30, label: '30 minutes before' },
  { value: 60, label: '1 hour before' },
  { value: 120, label: '2 hours before' },
  { value: 1440, label: '1 day before' },
] as const

/**
 * Type for reminder minute values
 */
export type ReminderMinutes = (typeof REMINDER_MINUTE_OPTIONS)[number]['value']

/**
 * Validation helpers
 */
export const NotificationSettingsValidation = {
  /**
   * Validate time format (HH:MM)
   * @param time - Time string to validate
   * @returns true if valid, false otherwise
   */
  isValidTimeFormat(time: string | null): boolean {
    if (!time) return false
    const timeRegex = /^([01]\d|2[0-3]):([0-5]\d)$/
    return timeRegex.test(time)
  },

  /**
   * Validate reminder minutes value
   * @param minutes - Minutes value to validate
   * @returns true if valid, false otherwise
   */
  isValidReminderMinutes(minutes: number | null): boolean {
    if (minutes === null) return false
    return REMINDER_MINUTE_OPTIONS.some((opt) => opt.value === minutes)
  },

  /**
   * Validate time format and return error message
   * @param time - Time string to validate
   * @returns Error message if invalid, null if valid
   */
  validateTimeFormat(time: string | null): string | null {
    if (!time) return 'Time is required'
    if (!this.isValidTimeFormat(time)) {
      return 'Time must be in HH:MM format (00:00 to 23:59)'
    }
    return null
  },

  /**
   * Validate reminder minutes and return error message
   * @param minutes - Minutes value to validate
   * @returns Error message if invalid, null if valid
   */
  validateReminderMinutes(minutes: number | null): string | null {
    if (minutes === null) return 'Reminder time is required'
    if (!this.isValidReminderMinutes(minutes)) {
      return 'Invalid reminder time. Please select from the available options.'
    }
    return null
  },
}

File: backend/src/pazpaz/schemas/notification_settings.py
"""Pydantic schemas for user notification settings API."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator
from pydantic_core import PydanticCustomError


class NotificationSettingsResponse(BaseModel):
    """Response schema for user notification settings.

    This schema represents the complete notification settings for a user,
    including all email notification preferences, digest settings, and reminders.
    """

    id: uuid.UUID
    user_id: uuid.UUID
    workspace_id: uuid.UUID

    # Master toggle
    email_enabled: bool = Field(
        ...,
        description="Master toggle - when false, no emails are sent",
    )

    # Event notifications
    notify_appointment_booked: bool = Field(
        ...,
        description="Send email when new appointment is booked",
    )
    notify_appointment_cancelled: bool = Field(
        ...,
        description="Send email when appointment is cancelled",
    )
    notify_appointment_rescheduled: bool = Field(
        ...,
        description="Send email when appointment is rescheduled",
    )
    notify_appointment_confirmed: bool = Field(
        ...,
        description="Send email when client confirms appointment",
    )

    # Daily digest settings
    digest_enabled: bool = Field(
        ...,
        description="Enable daily digest email (opt-in)",
    )
    digest_time: str | None = Field(
        None,
        description="Time to send digest in HH:MM format (24-hour, workspace timezone)",
        examples=["08:00", "18:30"],
    )
    digest_days: list[int] = Field(
        ...,
        min_length=1,
        description="Days of week to send digest (0=Sunday, 1=Monday, ..., 6=Saturday)",
        examples=[[1, 2, 3, 4, 5], [1, 3, 5], [0, 1, 2, 3, 4, 5, 6]],
    )

    # Appointment reminder settings
    reminder_enabled: bool = Field(
        ...,
        description="Enable appointment reminder emails",
    )
    reminder_minutes: int | None = Field(
        None,
        description="Minutes before appointment to send reminder (15, 30, 60, 120, 1440)",
        examples=[60, 1440],
    )

    # Session notes reminder settings
    notes_reminder_enabled: bool = Field(
        ...,
        description="Enable draft session notes reminders",
    )
    notes_reminder_time: str | None = Field(
        None,
        description="Time to send notes reminder in HH:MM format (24-hour, workspace timezone)",
        examples=["18:00", "20:00"],
    )

    # Future extensibility
    extended_settings: dict[str, Any] | None = Field(
        None,
        description="Future notification preferences (SMS, push, quiet hours, etc.)",
    )

    # Audit fields
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class NotificationSettingsUpdate(BaseModel):
    """Request schema for updating user notification settings.

    All fields are optional to support partial updates.
    Only provided fields will be updated.
    """

    # Master toggle
    email_enabled: bool | None = Field(
        None,
        description="Master toggle - when false, no emails are sent",
    )

    # Event notifications
    notify_appointment_booked: bool | None = Field(
        None,
        description="Send email when new appointment is booked",
    )
    notify_appointment_cancelled: bool | None = Field(
        None,
        description="Send email when appointment is cancelled",
    )
    notify_appointment_rescheduled: bool | None = Field(
        None,
        description="Send email when appointment is rescheduled",
    )
    notify_appointment_confirmed: bool | None = Field(
        None,
        description="Send email when client confirms appointment",
    )

    # Daily digest settings
    digest_enabled: bool | None = Field(
        None,
        description="Enable daily digest email (opt-in)",
    )
    digest_time: str | None = Field(
        None,
        description="Time to send digest in HH:MM format (24-hour, workspace timezone)",
        examples=["08:00", "18:30"],
    )
    digest_days: list[int] | None = Field(
        None,
        min_length=1,
        description="Days of week to send digest (0=Sunday, 1=Monday, ..., 6=Saturday)",
        examples=[[1, 2, 3, 4, 5], [1, 3, 5], [0, 1, 2, 3, 4, 5, 6]],
    )

    # Appointment reminder settings
    reminder_enabled: bool | None = Field(
        None,
        description="Enable appointment reminder emails",
    )
    reminder_minutes: int | None = Field(
        None,
        description="Minutes before appointment to send reminder (15, 30, 60, 120, 1440)",
        examples=[60, 1440],
    )

    # Session notes reminder settings
    notes_reminder_enabled: bool | None = Field(
        None,
        description="Enable draft session notes reminders",
    )
    notes_reminder_time: str | None = Field(
        None,
        description="Time to send notes reminder in HH:MM format (24-hour, workspace timezone)",
        examples=["18:00", "20:00"],
    )

    # Future extensibility
    extended_settings: dict[str, Any] | None = Field(
        None,
        description="Future notification preferences (SMS, push, quiet hours, etc.)",
    )

    @field_validator("digest_time", "notes_reminder_time")
    @classmethod
    def validate_time_format(cls, v: str | None) -> str | None:
        """
        Validate time format is HH:MM in 24-hour format.

        Args:
            v: Time string to validate

        Returns:
            Validated time string

        Raises:
            PydanticCustomError: If time format is invalid
        """
        if v is None:
            return v

        # Import validation from model
        from pazpaz.models.user_notification_settings import UserNotificationSettings

        if not UserNotificationSettings.is_valid_time_format(v):
            raise PydanticCustomError(
                "time_format",
                "Time must be in HH:MM format (00:00 to 23:59), got: {time}",
                {"time": v},
            )

        return v

    @field_validator("reminder_minutes")
    @classmethod
    def validate_reminder_minutes(cls, v: int | None) -> int | None:
        """
        Validate reminder minutes is one of the allowed presets.

        Args:
            v: Minutes value to validate

        Returns:
            Validated minutes value

        Raises:
            PydanticCustomError: If minutes is not an allowed preset
        """
        if v is None:
            return v

        # Import validation from model
        from pazpaz.models.user_notification_settings import UserNotificationSettings

        if not UserNotificationSettings.is_valid_reminder_minutes(v):
            raise PydanticCustomError(
                "reminder_minutes_invalid",
                "reminder_minutes must be one of: 15, 30, 60, 120, 1440, got: {minutes}",
                {"minutes": v},
            )

        return v

    @field_validator("digest_days")
    @classmethod
    def validate_digest_days(cls, v: list[int] | None) -> list[int] | None:
        """
        Validate digest_days array contains valid day numbers.

        Args:
            v: List of day numbers to validate (0=Sunday, 6=Saturday)

        Returns:
            Validated and normalized list of day numbers (sorted, deduplicated)

        Raises:
            PydanticCustomError: If any day number is outside 0-6 range or list is empty
        """
        if v is None:
            return v

        # Ensure at least one day is selected
        if len(v) == 0:
            raise PydanticCustomError(
                "digest_days_empty",
                "At least one day must be selected for digest",
                {},
            )

        # Validate all values are 0-6
        if not all(0 <= day <= 6 for day in v):
            invalid_days = [day for day in v if not 0 <= day <= 6]
            raise PydanticCustomError(
                "digest_days_invalid",
                "digest_days must contain values between 0 and 6, got invalid values: {invalid_days}",
                {"invalid_days": invalid_days},
            )

        # Remove duplicates and sort
        return sorted(list(set(v)))
